racket/collects/json/main.rkt

```diff
@@ -138,26 +155,47 @@
           [(list? x)
            (write-bytes #"[" o)
            (when (pair? x)
-             (loop (car x))
-             (for ([x (in-list (cdr x))]) (write-bytes #"," o) (loop x)))
+             (let ([layer (add1 layer)])
+               (let ([x (car x)])
+                 (format/write-newline)
+                 (format/write-indent layer)
+                 (write-jsval x layer))
+               (for ([x (in-list (cdr x))])
+                 (write-bytes #"," o)
+                 (format/write-newline)
+                 (format/write-indent layer)
+                 (write-jsval x layer)))
+             (format/write-newline)
+             (format/write-indent layer))
            (write-bytes #"]" o)]
           [(hash? x)
-           (write-bytes #"{" o)
            (define first? #t)
-           (hash-for-each
-            x
-            (lambda (k v)
-              (unless (symbol? k)
-                (raise-type-error who "legal JSON key value" k))
-              (if first? (set! first? #f) (write-bytes #"," o))
-              ;; use a string encoding so we get the same deal with
-              ;; `rx-to-encode'
-              (write-json-string (symbol->string k))
-              (write-bytes #":" o)
-              (loop v))
-            ;; order output
-            #true)
-            (write-bytes #"}" o)]
+           (define (write-hash-kv layer)
+             (λ (k v)
+               (unless (symbol? k)
+                 (raise-type-error who "legal JSON key value" k))
+               (if first? (set! first? #f) (write-bytes #"," o))
+               (format/write-newline)
+               (format/write-indent layer)
+               ;; use a string encoding so we get the same deal with
+               ;; `rx-to-encode'
+               (write-json-string (symbol->string k))
```

Comment:
Not necessary for this PR, but using [`symbol->immutable-string`](https://docs.racket-lang.org/reference/symbols.html#%28def._%28%28lib._racket%2Fsymbol..rkt%29._symbol-~3eimmutable-string%29%29) here could avoid some allocation.